File: CommonPython/spark_agg_methods_common_python/challenges/six_field_test_data/six_test_data_types.py
import logging
import os
from dataclasses import dataclass
from typing import Any, NamedTuple

import pandas as pd
import pyarrow

from spark_agg_methods_common_python.perf_test_common import (
    LOCAL_TEST_DATA_FILE_LOCATION, Challenge, ChallengeMethodRegistrationBase,
    DataSetDescriptionBase, ExecutionParametersBase,
    NumericalToleranceExpectations, TChallengeMethodDelegate,
    TSolutionInterface,
)

logger = logging.getLogger(__name__)

# cSpell: ignore arange


SIX_TEST_CHALLENGES = [Challenge.BI_LEVEL, Challenge.CONDITIONAL, Challenge.VANILLA]
TARGET_PARQUET_BATCH_SIZE = 65536


class DataPointNT(NamedTuple):
    id: int
    grp: int
    subgrp: int
    A: float
    B: float
    C: float
    D: float
    E: float
    F: float


SIX_TEST_SOURCE_DATA_PYARROW_SCHEMA = pyarrow.schema([
    pyarrow.field("id", pyarrow.int32()),
    pyarrow.field("grp", pyarrow.int32()),
    pyarrow.field("subgrp", pyarrow.int32()),
    pyarrow.field("A", pyarrow.int32()),
    pyarrow.field("B", pyarrow.int32()),
    pyarrow.field("C", pyarrow.float64()),
    pyarrow.field("D", pyarrow.float64()),
    pyarrow.field("E", pyarrow.float64()),
    pyarrow.field("F", pyarrow.float64()),
])


@dataclass(frozen=True)
class SubTotalDC():
    running_count: int
    running_max_of_D: float
    running_sum_of_C: float
    running_sum_of_E_squared: float
    running_sum_of_E: float


class SubTotalNT(NamedTuple):
    running_sum_of_C: float
    running_count: int
    running_max_of_D: float | None
    running_sum_of_E_squared: float
    running_sum_of_E: float


@dataclass(frozen=True)
class TotalDC():
    mean_of_C: float
    max_of_D: float
    var_of_E: float
    var_of_E2: float


class SixTestDataSetDescription(DataSetDescriptionBase):
    num_grp_1: int
    num_grp_2: int
    points_per_index: int
    relative_cardinality_between_groupings: int

    def __init__(
            self,
            *,
            debugging_only: bool,
            num_grp_1: int,
            num_grp_2: int,
            points_per_index: int,
            size_code: str,
    ):
        num_source_rows = num_grp_1 * num_grp_2 * points_per_index
        relative_cardinality_between_groupings = num_grp_2 // num_grp_1
        super().__init__(
            debugging_only=debugging_only,
            num_source_rows=num_source_rows,
            size_code=size_code,
        )
        self.num_grp_1 = num_grp_1
        self.num_grp_2 = num_grp_2
        self.points_per_index = points_per_index
        self.relative_cardinality_between_groupings = relative_cardinality_between_groupings


@dataclass(frozen=True)
class SixTestDataSetAnswers():
    bi_level_answer: pd.DataFrame
    conditional_answer: pd.DataFrame
    vanilla_answer: pd.DataFrame

    def answer_for_challenge(
            self,
            challenge: Challenge,
    ) -> pd.DataFrame:
        match challenge:
            case Challenge.VANILLA:
                return self.vanilla_answer
            case Challenge.BI_LEVEL:
                return self.bi_level_answer
            case Challenge.CONDITIONAL:
                return self.conditional_answer
            case _:
                raise KeyError(f"Unknown challenge {challenge}")


@dataclass(frozen=True)
class SixTestExecutionParameters(ExecutionParametersBase):
    pass


@dataclass(frozen=True)
class SixTestDataSetWAnswers():
    num_source_rows: int
    src_num_partitions: int
    tgt_num_partitions_1_level: int
    tgt_num_partitions_2_level: int
    df_src: pd.DataFrame
    vanilla_answer: pd.DataFrame
    bilevel_answer: pd.DataFrame
    conditional_answer: pd.DataFrame


@dataclass(frozen=True)
class SixTestDataChallengeMethodRegistrationBase(
    ChallengeMethodRegistrationBase[TSolutionInterface, TChallengeMethodDelegate]
):
    numerical_tolerance: NumericalToleranceExpectations


@dataclass(frozen=True)
class SixTestSourceDataFilePaths():
    source_directory_path: str
    source_file_path_parquet_small_v1_files: str
    source_file_path_parquet_single_file: str
    source_file_path_csv: str

    @property
    def file_paths(self) -> list[str]:
        return [
            self.source_file_path_parquet_small_v1_files,
            self.source_file_path_parquet_single_file,
            self.source_file_path_csv,
        ]


def six_derive_expected_answer_data_file_paths_csv(
        data_description: SixTestDataSetDescription,
        *,
        temp_file: bool = False,
) -> dict[Challenge, str]:
    num_grp_1 = data_description.num_grp_1
    num_grp_2 = data_description.num_grp_2
    repetition = data_description.points_per_index
    temp_postfix = "_temp" if temp_file else ""
    answer_file_names = {
        challenge: os.path.join(
            LOCAL_TEST_DATA_FILE_LOCATION,
            "SixField_Test_Data",
            f"{challenge}_answer_data_{num_grp_1}_{num_grp_2}_{repetition}{temp_postfix}.csv")
        for challenge in SIX_TEST_CHALLENGES
    }
    return answer_file_names


def six_derive_source_test_data_file_path(
        data_description: SixTestDataSetDescription,
        *,
        temp_file: bool = False,
) -> SixTestSourceDataFilePaths:
    num_grp_1 = data_description.num_grp_1
    num_grp_2 = data_description.num_grp_2
    repetition = data_description.points_per_index
    temp_postfix = "_temp" if temp_file else ""
    source_directory_path = os.path.join(
        LOCAL_TEST_DATA_FILE_LOCATION,
        "SixField_Test_Data",
    )
    stem = os.path.join(
        source_directory_path,
        f"six_field_source_data_{num_grp_1}_{num_grp_2}_{repetition}{temp_postfix}",
    )
    return SixTestSourceDataFilePaths(
        source_directory_path=source_directory_path,
        source_file_path_parquet_small_v1_files=f"{stem}_spark",
        source_file_path_parquet_single_file=f"{stem}_modern.parquet",
        source_file_path_csv=f"{stem}.csv",
    )


def fetch_six_data_set_answer(
        challenge: Challenge,
        data_size: SixTestDataSetDescription,
        *,
        spark_logger: Any | None = None,
) -> pd.DataFrame:
    answer_file_path_csv = six_derive_expected_answer_data_file_paths_csv(data_size)[challenge]
    (spark_logger or logger).info(f"Loading answer from {answer_file_path_csv}")
    return pd.read_csv(answer_file_path_csv)

File: CSharp/CSharpDataFrame/Challenges/SixFieldTestData/SixTestDataTypes.cs
namespace CSharpDataFrame.Challenges.SixFieldTestData.SixTestDataTypes;

using System.Collections.Generic;
using System.IO;
using CSharpDataFrame.SharedBetweenChallenges.CommonTypes;
using Microsoft.Data.Analysis;

public static class SixFieldTestDataConstants
{
    public static readonly Challenge[] SIX_TEST_CHALLENGES = new Challenge[]
    {
        Challenge.BI_LEVEL,
        Challenge.CONDITIONAL,
        Challenge.VANILLA
    };
    public const int TARGET_PARQUET_BATCH_SIZE = 65536;
}

public record DataPoint
{
    public int Id { get; init; }
    public int Grp { get; init; }
    public int SubGrp { get; init; }
    public float A { get; init; }
    public float B { get; init; }
    public float C { get; init; }
    public float D { get; init; }
    public float E { get; init; }
    public float F { get; init; }
}

public record SubTotal
{
    public int RunningCount { get; init; }
    public double RunningMaxOfD { get; init; }
    public double RunningSumOfC { get; init; }
    public double RunningSumOfESquared { get; init; }
    public double RunningSumOfE { get; init; }
}


public record TotalDC(
double MeanOfC,
double MaxOfD,
double VarOfE,
double VarOfE2
);


public abstract class SixTestDataSetDescription : DataSetDescriptionBase
{
    public int NumGrp1 { get; init; }
    public int NumGrp2 { get; init; }
    public int PointsPerIndex { get; init; }
    public int RelativeCardinalityBetweenGroups => NumGrp2 / NumGrp1;

    public SixTestDataSetDescription(
    bool debuggingOnly, int numGrp1, int numGrp2, int pointsPerIndex, string sizeCode
    ) : base(
    debuggingOnly: debuggingOnly,
    numSourceRows: numGrp1 * numGrp2 * pointsPerIndex,
    sizeCode: sizeCode
    )
    { }


    public static Dictionary<Challenge, string> SixDeriveExpectedAnswerDataFilePathsCsv(
        SixTestDataSetDescription dataDescription,
        bool tempFile = false)
    {
        int numGrp1 = dataDescription.NumGrp1;
        int numGrp2 = dataDescription.NumGrp2;
        int repetition = dataDescription.PointsPerIndex;
        string tempPostfix = tempFile ? "_temp" : "";

        var answerFileNames = new Dictionary<Challenge, string>();
        foreach (var challenge in SixFieldTestDataConstants.SIX_TEST_CHALLENGES)
        {
            string filePath = Path.Combine(
                PerfTestCommonConstants.LOCAL_TEST_DATA_FILE_LOCATION,
                "SixField_Test_Data",
                $"{challenge}_answer_data_{numGrp1}_{numGrp2}_{repetition}{tempPostfix}.csv");
            answerFileNames[challenge] = filePath;
        }

        return answerFileNames;
    }


    public static SixTestSourceDataFilePaths SixDeriveSourceTestDataFilePath(
        SixTestDataSetDescription dataDescription,
        bool tempFile = false)
    {
        int numGrp1 = dataDescription.NumGrp1;
        int numGrp2 = dataDescription.NumGrp2;
        int repetition = dataDescription.PointsPerIndex;
        string tempPostfix = tempFile ? "_temp" : "";

        string sourceDirectoryPath = Path.Combine(
            PerfTestCommonConstants.LOCAL_TEST_DATA_FILE_LOCATION,
            "SixField_Test_Data"
        );

        string stem = Path.Combine(
            sourceDirectoryPath,
            $"six_field_source_data_{numGrp1}_{numGrp2}_{repetition}{tempPostfix}"
        );

        return new SixTestSourceDataFilePaths(
            sourceDirectoryPath,
            $"{stem}_spark",
            $"{stem}_modern.parquet",
            $"{stem}.csv"
        );
    }

}


public class SixTestDataSetAnswers
{
    public DataFrame BiLevelAnswer { get; }
    public DataFrame ConditionalAnswer { get; }
    public DataFrame VanillaAnswer { get; }

    public SixTestDataSetAnswers(DataFrame biLevelAnswer, DataFrame conditionalAnswer, DataFrame vanillaAnswer)
    {
        BiLevelAnswer = biLevelAnswer;
        ConditionalAnswer = conditionalAnswer;
        VanillaAnswer = vanillaAnswer;
    }

    public DataFrame AnswerForChallenge(Challenge challenge)
    {
        return challenge switch
        {
            Challenge.VANILLA => VanillaAnswer,
            Challenge.BI_LEVEL => BiLevelAnswer,
            Challenge.CONDITIONAL => ConditionalAnswer,
            _ => throw new KeyNotFoundException($"Unknown challenge {challenge}")
        };
    }
}

public record SixTestExecutionParameters(
    int defaultParallelism, int numExecutors
) : ExecutionParametersBase(defaultParallelism, numExecutors);


public record SixTestDataSetWAnswers(
    int NumSourceRows,
    int SrcNumPartitions,
    int TgtNumPartitions1Level,
    int TgtNumPartitions2Level,
    DataFrame DfSrc,
    DataFrame VanillaAnswer,
    DataFrame BilevelAnswer,
    DataFrame ConditionalAnswer
);


public class SixTestDataChallengeMethod : ChallengeMethodBase
{
    public double NumericalTolerance { get; init; }

    public SixTestDataChallengeMethod(
        string strategyName,
        SolutionLanguage language,
        CalcEngine engine,
        CSharpSolutionInterface solutionInterface,
        bool requiresGpu,
        double NumericalTolerance
    ) : base(strategyName, language, engine, solutionInterface, requiresGpu)
    {
        this.NumericalTolerance = NumericalTolerance;
    }
}

public class SixTestSourceDataFilePaths
{
    public string SourceDirectoryPath { get; }
    public string SourceFilePathParquetSmallV1Files { get; }
    public string SourceFilePathParquetSingleFile { get; }
    public string SourceFilePathCsv { get; }

    public SixTestSourceDataFilePaths(
        string sourceDirectoryPath,
        string sourceFilePathParquetSmallV1Files,
        string sourceFilePathParquetSingleFile,
        string sourceFilePathCsv)
    {
        SourceDirectoryPath = sourceDirectoryPath;
        SourceFilePathParquetSmallV1Files = sourceFilePathParquetSmallV1Files;
        SourceFilePathParquetSingleFile = sourceFilePathParquetSingleFile;
        SourceFilePathCsv = sourceFilePathCsv;
    }

    public string[] FilePaths => [
        SourceFilePathParquetSmallV1Files,
        SourceFilePathParquetSingleFile,
        SourceFilePathCsv
    ];
}

